text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,267 @@
+
+```html
+<template for:each={items} for:item="item">
+  <div key={item.id}></div>
+  <span key={item.id}></span>
+  <button key={item.id}></button>
+</template>
+```
+
+With the new directives, this becomes:
+
+```html
+<template lwc:each={items} lwc:item="item" lwc:key={item.id}>
+  <div></div>
+  <span></span>
+  <button></button>
+</template>
+```
+
+Also note that, in the shorthand `<template>`-less format, `lwc:ref` is defined not on a `<template>` but on the same element as the `lwc:each`:
+
+```html
+<li lwc:each={items} lwc:item="item" lwc:key={item.id}>
+    {item.name}
+</li>
+```
+
+Similar to `key`, `lwc:key` is required on elements with `lwc:each`, and throws an error if missing:
+
+    Missing lwc:key for lwc:each. Iterators must have a unique, computed key value.
+
+Unlike `key`, a missing `lwc:key` throws the error regardless of the content inside the iterator. (A missing `key` will not throw for text nodes, comment nodes, or empty markup – an apparent [inconsistency](https://github.com/salesforce/lwc/issues/3860) in that directive.)
+
+Similar to `key`, `lwc:key` must not reference the variable defined by `lwc:index`. It also must not reference the variable defined by `lwc:first`, `lwc:last`, `lwc:even`, or `lwc:odd`. For example, this will throw a compile-time error:
+
+```html
+<template lwc:each={items} lwc:item="item" lwc:index="idx" lwc:key={idx}>
+</template>
+```
+
+In short, `lwc:key` _must_ reference the `lwc:item` variable. Anything else throws a compile-time error.
+
+### `lwc:index`
+
+Identical to `for:index`. Returns a number for the current list index.
+
+As with `for:index`, `lwc:index` cannot use the same string as the `lwc:item` on the same element:
+
+```html
+<template lwc:each={items} lwc:key={same.id} lwc:item="same" lwc:index="same">
+</template>
+```
+
+The above will throw a compile-time error.
+
+(Note that `for:index` only [incidentally](https://stackblitz.com/edit/salesforce-lwc-se51rr?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground) throws an error for the same case, because the `key` ends up referencing the `for:index` variable. For `lwc:index`, the compile-time error should be explicit.)
+
+### `lwc:first`, `lwc:last`, `lwc:even`, `lwc:odd`
+
+These new directives take strings as their values. When referenced, they resolve to a boolean representing whether the list item is first, last, even, or odd. (This is heavily inspired by [Angular](https://angular.io/api/common/NgForOf#local-variables).)
+
+The addition of `lwc:first` and `lwc:last` serve to unify the two existing directives – previously, `iterator:*` was the only way to get "first" or "last" behavior. The addition of `lwc:even` and `lwc:odd` are merely for convenience (especially in the absence of [complex template expressions](https://github.com/salesforce/lwc/pull/3376)).
+

```

Comment:
In a nested iteration, can the string values for these variables have to be unique? In other words, if an outer and inner `lwc:each` share the same variable, will it result in a uniqueness constraint violation or will it allow it and apply  
lexical scoping.

``` 
       <template lwc:each={contacts} 
              lwc:item="contact"
              lwc:key={contact.id}
              lwc:first="first">
             <template lwc:each={contact.addresses} 
                   lwc:item="address"
                   lwc:key={address.id}
                   lwc:first="first">
            </template>
     </template>
```

Reply:
It looks like the current behavior of `for:each` is that [inner loops shadow the outer loops](https://stackblitz.com/edit/salesforce-lwc-6usqmw?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground). In other words, inner loops can reuse the same variable names as outer loops, but inside of those inner loops, the inner loop variables override the outer loop variables.

I think this behavior makes sense, since it mirrors how variable shadowing works in JS itself:

```js
function outer() {
  const foo = 'bar'
  function inner() {
    const foo = 'quux' // shadows the outer variable
  }
}
```

I can add some language to the RFC to clarify this.

Reply:
2195f72